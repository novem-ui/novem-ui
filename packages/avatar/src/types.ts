export type AvatarSizes = 'sm' | 'md' | 'lg'

export interface AvatarProps {
  description?: string
  image?: string
  name?: string
  size?: AvatarSizes
}
