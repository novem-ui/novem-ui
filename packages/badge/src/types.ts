import theme from '@novem-ui/theme'
import { Hierarchy } from '@novem-ui/base'
import { Theme } from '@emotion/react'

export type BadgeHierarchy = Exclude<Hierarchy, 'link'>
export type BadgeSize = 'sm' | 'lg'

export interface BadgeProps {
  hierarchy?: BadgeHierarchy
  size?: BadgeSize
  baseColor?: keyof Omit<Omit<typeof theme.colors.palette, 'base'>, 'darkGrey'>
}

export type BadgeUtilsFunction = (props: BadgeProps & { theme?: Theme }) => string
