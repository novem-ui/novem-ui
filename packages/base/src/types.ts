import { Property } from 'csstype'
import { BaseColor, Color, SpacingValue } from '@novem-ui/theme'
import { Theme } from '@emotion/react'

// Spacing types
export type SpacingPropName =
  | 'm'
  | 'mt'
  | 'mr'
  | 'mb'
  | 'ml'
  | 'mx'
  | 'my'
  | 'margin'
  | 'marginX'
  | 'marginY'
  | 'marginTop'
  | 'marginRight'
  | 'marginBottom'
  | 'marginLeft'
  | 'p'
  | 'pt'
  | 'pr'
  | 'pb'
  | 'pl'
  | 'px'
  | 'py'
  | 'padding'
  | 'paddingX'
  | 'paddingY'
  | 'paddingTop'
  | 'paddingRight'
  | 'paddingBottom'
  | 'paddingLeft'

export type SpacingPropValue = SpacingValue | string

export type SpacingPropEntry = [SpacingPropName, SpacingPropValue]

export type SpacingProps = { [key in SpacingPropName]?: SpacingPropValue }

export type CSSSpacingProps = Omit<
  SpacingProps,
  'm' | 'mt' | 'mr' | 'mb' | 'ml' | 'mx' | 'my' | 'p' | 'pt' | 'pr' | 'pb' | 'pl' | 'px' | 'py'
>

// Text types
export interface TypographyProps<As, Size> {
  align?: Property.TextAlign
  as?: As
  color?: Color | Property.Color
  decoration?: Property.TextDecoration
  direction?: Property.Direction
  family?: Property.FontFamily
  kerning?: Property.FontKerning
  letterSpacing?: Property.LetterSpacing
  lineHeight?: Property.LineHeight
  orientation?: Property.TextOrientation
  transform?: Property.TextTransform
  size?: Size | Property.FontSize
  shadow?: Property.TextShadow
  stretch?: Property.FontStretch
  fontStyle?: Property.FontStyle
  theme?: Theme
  variantCaps?: Property.FontVariantCaps
  variantNumeric?: Property.FontVariantNumeric
  weight?: Property.FontWeight
  whiteSpace?: Property.FontWeight
  wordBreak?: Property.WordBreak
  wordSpacing?: Property.WordSpacing
  writingMode?: Property.WritingMode
}

export type TextComponentType = 'heading' | 'text'
export type HeadingTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

// Hierarchy
export type Hierarchy = 'loud' | 'quiet' | 'link'

// InputElement
export interface InputElementProps {
  theme?: Theme
  baseColor?: BaseColor
  error?: boolean | string | Error
  valid?: boolean
  disabled?: boolean
  readOnly?: boolean
}

// Layout

export interface LayoutProps {
  display?: Property.Display
  position?: Property.Position
  top?: Property.Top
  bottom?: Property.Bottom
  height?: Property.Height
  left?: Property.Left
  right?: Property.Right
  width?: Property.Width
}
