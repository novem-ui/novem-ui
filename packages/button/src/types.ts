import { Hierarchy, SpacingProps } from '@novem-ui/base'
import theme from '@novem-ui/theme'

export type ButtonSize = 'sm' | 'md' | 'lg'

export interface ButtonProps extends SpacingProps{
  hierarchy?: Hierarchy
  size?: ButtonSize
  baseColor?: keyof typeof theme.colors.palette
  disabled?: boolean
}
