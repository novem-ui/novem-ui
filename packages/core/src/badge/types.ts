import * as React from 'react'

export default interface IBadge {
  variant: string,
  children: React.ReactElement,
}
