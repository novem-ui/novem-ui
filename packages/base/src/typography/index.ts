import { css } from '@emotion/react'
import { Theme } from '@novem-ui/theme'
import { TypographyProps } from '../types'

type TextComponentType = 'heading' | 'text'
export type HeadingTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

function getHeadingFontSize(level: HeadingTags = 'h1', sizes: Theme['typography']['headingFontSizes']) {
  const mappedSizes = {
    h1: sizes.xxl,
    h2: sizes.xl,
    h3: sizes.lg,
    h4: sizes.md,
    h5: sizes.sm,
    h6: sizes.xs,
  }

  return mappedSizes[level]
}

function typography<As = string, Size = string>(
  {
    align,
    as,
    color,
    decoration,
    direction,
    family,
    kerning,
    letterSpacing,
    lineHeight,
    orientation,
    transform,
    size,
    shadow,
    stretch,
    style,
    theme,
    variantCaps,
    variantNumeric,
    weight,
    whiteSpace,
    wordBreak,
    wordSpacing,
    writingMode,
  }: TypographyProps<As, Size> & { theme: Theme },
  component: TextComponentType
) {
  const isHeading = component === 'heading'
  const { typography } = theme
  const fontSizes = typography[isHeading ? 'headingFontSizes' : 'textFontSizes']
  const fontSize = (fontSizes[size as keyof typeof fontSizes]
    ? `${fontSizes[size as keyof typeof fontSizes]}px`
    : size) as string
  const defaultFontSize = isHeading
    ? `${getHeadingFontSize(as as any, fontSizes as Theme['typography']['headingFontSizes'])}px`
    : `${fontSizes.md}px`

  const styles = css`
    color: ${color};
    font-size: ${fontSize || defaultFontSize};
    font-weight: ${weight};
    margin: 0;

    ${align && `text-align: ${align};`}
    ${decoration && `text-decoration: ${decoration};`}
    ${direction && `text-direction: ${direction};`}
    ${family && `font-family: ${family};`}
    ${kerning && `font-kerning: ${kerning};`}
    ${letterSpacing && `letter-spacing: ${letterSpacing};`}
    ${lineHeight && `line-height: ${lineHeight};`}
    ${orientation && `text-orientation: ${orientation};`}
    ${transform && `text-transform: ${transform};`}
    ${shadow && `text-shadow: ${shadow};`}
    ${stretch && `:font-stretch ${stretch};`}
    ${style && `font-style: ${style};`}
    ${variantCaps && `font-variant-caps: ${variantCaps};`}
    ${variantNumeric && `font-variant-numeric: ${variantNumeric};`}
    ${whiteSpace && `whiteSpace: ${whiteSpace};`}
    ${wordBreak && `word-break: ${wordBreak};`}
    ${wordSpacing && `word-spacing: ${wordSpacing};`}
    ${writingMode && `writing-mode: ${writingMode};`}
  `

  return styles
}

export default typography
