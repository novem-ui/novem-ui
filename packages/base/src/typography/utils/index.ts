import { Theme } from '@emotion/react'
import get from 'lodash.get'
import { HeadingTags, TextComponentType, TypographyProps } from '../../types'

export interface GetFontSizesConfig {
  component: TextComponentType
  theme: Theme
}

export function getHeadingFontSize(level: HeadingTags = 'h1', sizes: Theme['typography']['headingFontSizes']) {
  const mappedSizes = {
    h1: sizes.xxl,
    h2: sizes.xl,
    h3: sizes.lg,
    h4: sizes.md,
    h5: sizes.sm,
    h6: sizes.xs
  }

  return mappedSizes[level]
}

export function getFontSizes<As, Size>({
  size,
  component,
  theme,
  as
}: Pick<TypographyProps<As, Size>, 'size' | 'as'> & GetFontSizesConfig) {
  const isHeading = component === 'heading'
  const { typography } = theme
  const fontSizes = typography[isHeading ? 'headingFontSizes' : 'textFontSizes']
  const fontSize = (fontSizes[size as keyof typeof fontSizes]
    ? `${fontSizes[size as keyof typeof fontSizes]}px`
    : size) as string
  const defaultFontSize = isHeading
    ? `${getHeadingFontSize(as as any, fontSizes as Theme['typography']['headingFontSizes'])}px`
    : `${fontSizes.md}px`

  return { fontSize, defaultFontSize }
}

export function getColor({ theme, color }: TypographyProps<string, string>) {
  if (color === 'theme') {
    return theme.colors.palette[theme.colors.base]['500']
  }

  const colorValue = get(theme.colors.palette, color)

  return colorValue || color
}
