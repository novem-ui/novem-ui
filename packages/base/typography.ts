export { default as typography, HeadingTags } from './src/typography'
export { TypographyProps } from './src/types'
