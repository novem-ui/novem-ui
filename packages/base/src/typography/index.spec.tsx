import React from 'react'

import { render } from '@testing-library/react'
import styled from '@emotion/styled'
import theme, { Theme } from '@novem-ui/theme'

import typography from '.'
import { TypographyProps } from '../types'

describe('typography', () => {
  test('should create a css fragment with the mapped props (partial)', () => {
    type Sizes = keyof Theme['typography']['headingFontSizes']
    type HeadingProps = TypographyProps<'h1', Sizes>

    const ComponentWithTypographyProps = styled.h1<HeadingProps>`
      ${(props) => typography<'h1', Sizes>(props, 'heading')}
    `
    const { container } = render(
      <ComponentWithTypographyProps theme={theme} color="red.500" family="sans-serif">
        test
      </ComponentWithTypographyProps>
    )

    expect(container).toMatchSnapshot()
  })

  test('should create a css fragment with the mapped props (partial with size)', () => {
    type ParagraphProps = TypographyProps<'p', keyof Theme['typography']['textFontSizes']>

    const ComponentWithTypographyProps = styled.p<ParagraphProps>`
      ${(props) => typography<'p', keyof Theme['typography']['textFontSizes']>(props, 'text')}
    `
    const { container } = render(
      <ComponentWithTypographyProps theme={theme} color="red.500" family="sans-serif" size="sm">
        test
      </ComponentWithTypographyProps>
    )

    expect(container).toMatchSnapshot()
  })

  test('should create a css fragment with the mapped props', () => {
    type ParagraphProps = TypographyProps<'p' | 'b', keyof Theme['typography']['textFontSizes']>

    const props: ParagraphProps = {
      align: 'right',
      as: 'b',
      color: 'yellow.500',
      decoration: 'dotted',
      direction: 'rtl',
      family: 'monospace',
      kerning: 'auto',
      letterSpacing: 'normal',
      lineHeight: 2,
      orientation: 'revert',
      transform: 'uppercase',
      size: 'xs',
      shadow: '1px 1px 2px black',
      stretch: 'extra-expanded',
      fontStyle: 'italic',
      variantCaps: 'small-caps',
      variantNumeric: 'slashed-zero',
      weight: 'bolder',
      whiteSpace: 'normal',
      wordBreak: 'break-word',
      wordSpacing: 'unset',
      writingMode: 'sideways-lr'
    }

    const ComponentWithTypographyProps = styled.p<ParagraphProps>`
      ${(props) => typography<'p' | 'b', keyof Theme['typography']['textFontSizes']>(props, 'text')}
    `
    const { container } = render(
      <ComponentWithTypographyProps theme={theme} {...props}>
        test
      </ComponentWithTypographyProps>
    )

    expect(container).toMatchSnapshot()
  })
})
