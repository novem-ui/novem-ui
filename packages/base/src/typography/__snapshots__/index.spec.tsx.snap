// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typography should create a css fragment with the mapped props (partial with size) 1`] = `
.emotion-0 {
  color: #EB5757;
  font-size: 14px;
  margin: 0;
  font-family: sans-serif;
}

<div>
  <p
    class="emotion-0 emotion-1"
    color="red.500"
  >
    test
  </p>
</div>
`;

exports[`typography should create a css fragment with the mapped props (partial) 1`] = `
.emotion-0 {
  color: #EB5757;
  font-size: 56px;
  margin: 0;
  font-family: sans-serif;
}

<div>
  <h1
    class="emotion-0 emotion-1"
    color="red.500"
  >
    test
  </h1>
</div>
`;

exports[`typography should create a css fragment with the mapped props 1`] = `
.emotion-0 {
  color: #F1694B;
  font-size: 12px;
  font-weight: bolder;
  margin: 0;
  text-align: right;
  -webkit-text-decoration: dotted;
  text-decoration: dotted;
  text-direction: rtl;
  font-family: monospace;
  font-kerning: auto;
  letter-spacing: normal;
  line-height: 2;
  -webkit-text-orientation: revert;
  -moz-text-orientation: revert;
  -ms-text-orientation: revert;
  text-orientation: revert;
  text-transform: uppercase;
  text-shadow: 1px 1px 2px black;
  : font-stretchextra-expanded;
  font-style: italic;
  font-variant-caps: small-caps;
  font-variant-numeric: slashed-zero;
  whiteSpace: normal;
  word-break: break-word;
  word-spacing: unset;
  -webkit-writing-mode: sideways-lr;
  -ms-writing-mode: tb;
  writing-mode: sideways-lr;
}

<div>
  <b
    class="emotion-0 emotion-1"
    color="theme"
    direction="rtl"
    font-style="italic"
    kerning="auto"
    letter-spacing="normal"
    orientation="revert"
    transform="uppercase"
    word-spacing="unset"
    writing-mode="sideways-lr"
  >
    test
  </b>
</div>
`;
